summingbird-storm/src/main/scala/com/twitter/summingbird/storm/StormTopologyBuilder.scala

```diff
@@ -1,102 +1,238 @@
 import com.twitter.summingbird.option.JobId
 import com.twitter.summingbird.planner.{ Dag, FlatMapNode, SourceNode, SummerNode }
 import com.twitter.summingbird.storm.Constants._
-import com.twitter.summingbird.storm.builder.Topology.ReceivingId
-import com.twitter.summingbird.storm.builder.{ EdgeType, Topology }
+import com.twitter.summingbird.storm.builder.Topology
 import com.twitter.summingbird.storm.planner.StormNode
 import org.apache.storm.generated.StormTopology
 import org.slf4j.LoggerFactory
 import scala.reflect.ClassTag
-
 import scala.collection.{ Map => CMap }
 
-object StormTopologyBuilder {
+/**
+ * This class contains main logic on how to build storm topology out of planned DAG.
+ * `DAG` contains three types of nodes: Source, FlatMap and Summer.
+ * Built topology has next properties:
+ * 1. separate `Topology` `Component`s corresponds to separate DAG nodes
+ * 2. each `Component` emit values only of one type
+ * 3.a. `FlatMap` components accept tuples of type `Item[T]`
+ *      (and corresponds to `Producer[Storm, T]`)
+ * 3.b. `Summer` components accept tuples of type `SummerInput[K, V]`
+ *      (and corresponds to `KeyedProducer[Storm, K, V]`)
+ * 4. All components emit some of those: `Item[T]`, `Aggregated[K, V]` or `Sharded[K, V]`.
+ * 4.a. If corresponding to `Component` `Producer` is not `Keyed` then `Component` emits `Item[T]`
+ * 4.b. Otherwise if `Component` emits only to `Summer` nodes it emits `Aggregated[K, V]`
+ * 4.c. Otherwise if `Component` emits to both `Summer` and non `Summer` nodes it emits `Sharded[K, V]`
+ *      which can be both grouped in the same way as `Aggregated` and shuffled in the same way as `Item[T]`
+ */
+private[storm] object StormTopologyBuilder {
   @transient private val logger = LoggerFactory.getLogger(classOf[StormTopologyBuilder])
 
   type Item[T] = (Timestamp, T)
 
   type AggregateKey[K] = (K, BatchID)
   type AggregateValue[V] = (Timestamp, V)
   type Aggregated[K, V] = (Int, CMap[AggregateKey[K], AggregateValue[V]])
-
-  type ItemSpout[T] = Topology.Spout[Item[T]]
-  type AggregatedSpout[K, V] = Topology.Spout[Aggregated[K, V]]
-
-  type ItemFMBolt[T, U] = Topology.Bolt[Item[T], Item[U]]
-  type AggregatedFMBolt[T, K, V] = Topology.Bolt[Item[T], Aggregated[K, V]]
-
-  type SummerOutput[K, V] = Item[(K, (Option[V], V))]
-  type SummerBolt[K, V] = Topology.Bolt[Aggregated[K, V], SummerOutput[K, V]]
+  type Sharded[K, V] = (Int, AggregateKey[K], AggregateValue[V])
+
+  type SummerInput[K, V] = Iterable[(AggregateKey[K], AggregateValue[V])]
+
+  type FlatMapBoltId[T] = Topology.BoltId[Item[T], _]
+  type SummerBoltId[K, V] = Topology.BoltId[SummerInput[K, V], _]
+
+  private case class OutgoingSummersProps(
+    allSummers: Boolean,
+    batcher: Batcher,
+    semigroup: Semigroup[_],
+    shards: KeyValueShards
+  )
+
+  private def wrapSharded[K, V](
+    batcher: Batcher,
+    shards: KeyValueShards
+  ): (Item[(K, V)] => Sharded[K, V]) = { tuple =>
```

Comment:
Can we destructure tuple to improve readability:
  ): (Item[(K, V)] => Sharded[K, V]) = { case(ts, (k, v)) =>
    val wrappedKey = (k, batcher.batchOf(ts))
    val wrappedValue = (ts, v)
    val shardId = shards.summerIdFor(wrappedKey)
    (shardId, wrappedKey, wrappedValue)